diff/order.go

```diff
@@ -0,0 +1,84 @@
+package diff
+
+import "github.com/kong/deck/types"
+
+/*
+                                       Root
+                                         |
+         +----------+----------+---------+------------+---------------+
+         |          |          |         |            |               |
+         v          v          v         v            v               v
+L1    Service    RbacRole  Upstream  Certificate  CACertificate  Consumer ---+
+      Package        |         |        |     |      |                |      |
+        |            v         v        v     |      v                v      |
+L2      |        RBACRole   Target     SNI    +-> Service       Credentials  |
+        |        Endpoint                         |  |              (7)      |
+        |                                         |  |                       |
+        |                                         |  |                       |
+L3      +---------------------------> Service <---+  +-> Route               |
+        |                             Version        |     |                 |
+        |                                 |          |     |                 |
+        |                                 |          |     v                 |
+L4      +----------> Document   <---------+          +-> Plugins  <----------+
+*/
```

Comment:
Note to reviewers, please make sure that this dependency graph is accurate. The entire patch has been built based upon this.

Reply:
I like the gragh which brings in whole picture and clarity -- the message flow among dependencies. It would help to  understand if you add some concrete resource and actions in the picture. 

i.e. root (user) --> Certificate (Private, Public) --> SNI (authentication)

Reply:
This looks accurate, but was it written (drawn?) by hand? That seems like it'd be quite a pain to update--I already have nightmares enough from editing big Markdown tables by hand, and this is more complex :)

IMO we'd want to use some sort of metadata format that we can then generate an SVG from for a visual representation. There are several standard graph formats that we can use an off-the-shelf SVG generator for. https://networkx.org/documentation/stable/reference/readwrite/index.html covers most of them (along with NetworkX's proprietary formats).

Only issue there is that the standard formats are all XML, which isn't my favorite for writing by hand (though it's still better than hand-formatted ASCII). However, [GEXF](https://gephi.org/gexf/format/data.html) isn't particularly complex XML, and once we have an initial version available adding additional relationships should be easy, since you'd just copy-pasta node and edge entries and edit the contents.

To Jimin's point, I don't know that we'd need much in terms of edge labels, since these all just reflect foreign key relationships in the Kong DB schema rather than methods that interact with the other objects. The only one that's a bit non-obvious is the Certificate->Service relationship, which exists because of the client_certificate field. GEXF does support edge labels if we want those though (the example doesn't show them, but they are in the spec).

Reply:
Have you tried mermaid, [for example](https://mermaid.cloudinary.us/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbkFbUm9vdF0gLS4tPiBCW1NlcnZpY2UgUGFja2FnZV0gXG5BIC0uLT4gQ1tSYmFjUm9sZV1cbkEgLS4tPiBEW1Vwc3RyZWFtXVxuQSAtLi0-IEVbQ2VydGlmaWNhdGVdXG5BIC0uLT4gRltDQUNlcnRpZmljYXRlXVxuQSAtLi0-IEdbQ29uc3VtZXJdXG5DIC0uLT4gSFtSYmFjUm9sZUVuZHBvaW50XVxuRCAtLi0-IElbVGFyZ2V0XVxuRSAtLi0-IEpbU05JXVxuRSAtLi0-IEtbU2VydmljZV1cbkYgLS4tPiBLXG5HIC0uLT4gTFtDcmVkZW50aWFsc11cbkIgLS4tPiBNW1NlcnZpY2VWZXJzaW9uXVxuSyAtLi0-IE1cbksgLS4tPiBOW1JvdXRlXVxuQiAtLi0-IE9bRG9jdW1lbnRdXG5NIC0uLT4gT1xuSyAtLi0-IFBbUGx1Z2luXVxuTiAtLi0-IFBcbkcgLS4tPiBQIiwibWVybWFpZCI6eyJ0aGVtZSI6ImRlZmF1bHQifX0)?

Reply:
> This looks accurate, but was it written (drawn?) by hand? That seems like it'd be quite a pain to update--I already have nightmares enough from editing big Markdown tables by hand, and this is more complex :)

I used https://asciiflow.com to generate this diagram.

Reply:
>  It would help to understand if you add some concrete resource and actions in the picture.

All of these are entities in Kong and Konnect's data-model and nothing here is abstract.
I could be misunderstanding you, in which case please elaborate on what you mean.

Reply:
> I used https://asciiflow.com to generate this diagram.

Did you happen to save the share link for it? It looks like it can save the drawing using some custom URL data format, but doesn't have any way to import existing stuff.